Splitting concepts responsibility

FranckSix opened the pull request:
Closes #6.

Separated the concepts and responsibilities of the Code Generators to avoid the confusion caused by partial classes.

I applied RawString to format the unit manager code. This makes the code more readable and eliminates the use of AppendLineLF.

I reorganized the namespaces to focus efforts on the code generators and avoid the distractions of utility classes. All the business logic is now under Generators.

Utilities and satellite classes are under Tools.

Extension methods under Extensions.

Diagnostics have been grouped under the Rules class. I added a new one to track fatal errors and list it in the output.

I also added various extensions to streamline the code.

Reworking the code, I noticed that by not using the CodeGeneration option. The different classes don't go through the CurrentUICulture (which is problematic if you're using the tool in a Roslyn project).

Now that the class responsibilities are clearly defined, I could add a new Generator to the IStringLocalization model. This interface is used in WinUI and Blazor projects, among others.

FranckSix committed:
Splitting concepts responsibility

Separated the concepts and responsibilities of the Code Generators to avoid the confusion caused by partial classes.

I applied RawString to format the unit manager code. This makes the code more readable and eliminates the use of AppendLineLF.

I reorganized the namespaces to focus efforts on the code generators and avoid the distractions of utility classes. All the business logic is now under Generators.

Utilities and satellite classes are under Tools.

Extension methods under Extensions.

Diagnostics have been grouped under the Rules class. I added a new one to track fatal errors and list it in the output.

I also added various extensions to streamline the code.

Reworking the code, I noticed that by not using the CodeGeneration option. The different classes don't go through the CurrentUICulture (which is problematic if you're using the tool in a Roslyn project).

Now that the class responsibilities are clearly defined, I could add a new Generator to the IStringLocalization model. This interface is used in WinUI and Blazor projects, among others.

FranckSix committed:
Added compatibility with IStringLocalizer

Using heritage pattern over the generators to clarify witch every class do. To clarify and simplify concept, separation of concepts.

Unification of all Diagnostics in a class Analyzer to respect the Analyzer contract from MS RoslynAnalyser.

Rebrand the SourceGenerator class to switch over a new option GenerationType :
  - ResourceManager (Default)
  - CodeGeneration (In addidition to GenerateCode, to maintenability of existing code)
  - StringLocaliser (new generator includes registration classes to simplifing adding resources to the CollectionService) I added a demo project to demonstrate the new feature.

I optimized the uses of IncrementalProviders to avoid to apply the same filter over and over.

Now we have thee bloc of generation
- ResourcesManager
- CodeGeneration
   - ComboGeneration
- StringLocaliser
  - NamespaceRegistrator
  - GlobalRegistrator

ycanardeau commented on .editorconfig:
```diff
@@ -9,7 +9,7 @@ end_of_line = lf
 trim_trailing_whitespace = true
 insert_final_newline = true
-indent_style = tab
+indent_style = space
```
Can you please revert this change? Also, converting the indentation to tabs would make the review easier.

FranckSix replied:
Completed...

Sorry for inconvenience. I forgot to replace option before pull request. I work with spaces by habit. Indentation is more easy to manage  this way when working in a Team, or using multiple editors. I learned to use <Tab> indentation at school, then work team converted me to spaces at work 26 years ago 😲

I ended up loving it. It's worth a try. 😉

FranckSix committed:
Replace the <Tab> option

Replace the Tab in .editorconfig
readjust files to <tab> instead of spaces
readjust unit tests

ycanardeau commented on Aigamo.ResXGenerator/Tools/GenerationType.cs:
```diff
@@ -0,0 +1,9 @@
+﻿namespace Aigamo.ResXGenerator.Tools;
+
+public enum GenerationType
+{
+	ResourceManager,
+	CodeGeneration,
+	StringLocalizer,
+	SameAsOuter
```
Can you please add XML documentation comments for these enum values for better clarity?

FranckSix replied:
I report the README comment to XML Comment follow your recommendation in your further comment point

[d6cdf95](https://github.com/ycanardeau/ResXGenerator/pull/7/commits/d6cdf957ff6f310854fa7a68d47c825a22e68a6a)
[38397f2](https://github.com/ycanardeau/ResXGenerator/pull/7/commits/38397f2fb856dc5ddbeb377ce7227ed116e98844)

ycanardeau commented on Aigamo.ResXGenerator/Tools/GenFileOptions.cs:
```diff
@@ -17,11 +17,12 @@ public readonly record struct FileOptions
 	public string? CustomToolNamespace { get; init; }
 	public string LocalNamespace { get; init; }
 	public bool GenerateCode { get; init; }
+	public GenerationType GenerationType { get; init; }
```
This defaults to `ResourceManager`, but it's backward-compatible, right?

FranckSix replied:
Yes, my goal was to make it possible to add the IStringLocalizer while isolating each generation type. The choice to use ResourceManager by default reflects what ResXGenerator did before the refactor.

All my tests ensured that the generation without specific parameters produced the same result before and after.

FranckSix replied:
Also, even though you can now specify CodeGeneration, for compatibility reasons, the existing GenerateCode parameter is still evaluated.

GenerationType acts as a main switch for selecting the generation method. However, for ResourceManager, if the GenerateCode option is set to true, the ResXgenerator will still create the strongly-typed classes.

ycanardeau commented on Aigamo.ResXGenerator/Properties/launchSettings.json:
```diff
@@ -3,6 +3,9 @@
 			"commandName": "DebugRoslynComponent",
 			"targetProject": "..\\Aigamo.ResXGenerator.Tests\\Aigamo.ResXGenerator.Tests.csproj"
+		},
+		"Profil 1": {
+			"commandName": "Project"
 		}
```
What's the purpose of this profile?

FranckSix replied:
It seems to be added by VisualStudio. It can be ignored or rollback

ycanardeau commented on README.md:
```diff
@@ -16,60 +16,119 @@ Generated source from [ActivityEntrySortRuleNames.resx](https://github.com/VocaD
+
+public class MyViewModel(IArtistCategoriesNames resources) : IMyViewModel
+{
+	public string Value => resources.Actor // retur also the value string
```
There is a typo here.

FranckSix replied:
corrected [1632833](https://github.com/ycanardeau/ResXGenerator/pull/7/commits/1632833ad6ee0b1255386a7ed39bf468f8cae80c)

ycanardeau commented on README.md:
```diff
@@ -16,60 +16,119 @@ Generated source from [ActivityEntrySortRuleNames.resx](https://github.com/VocaD
+-	StringLocalizer
+	- When this option choosen the generator will generate interface and class to use with [Microsoft.Extensions.Localization](https://docs.microsoft.com/en-us/dotnet/core/extensions/localization) `IStringLocalizer<T>`. To see how to use it see [Using IStringLocalizer](#IStringLocalizer)
+-	SameAsOuter
+	- When this option choosen the generator will use the same generation type as the outer class if any. If no outer class exist it will fallback to 'ResourceManager'.
```
I think it should be fine to use these descriptions as XML documentation comments.

FranckSix replied:
Good point !

FranckSix replied:
I reported the README documentation in enum class, It's now available from intellisense. 🥰

[d6cdf95](https://github.com/ycanardeau/ResXGenerator/pull/7/commits/d6cdf957ff6f310854fa7a68d47c825a22e68a6a)
[38397f2](https://github.com/ycanardeau/ResXGenerator/pull/7/commits/38397f2fb856dc5ddbeb377ce7227ed116e98844)

ycanardeau commented on Aigamo.ResXGenerator/Tools/GroupedAdditionalFile.cs:
```diff
@@ -11,28 +13,17 @@ public GroupedAdditionalFile(AdditionalTextWithHash mainFile, IReadOnlyList<Addi
-				hashCode = (hashCode * 397) ^ additionalText.GetHashCode();
-			}
-
+			SubFiles.ForEach(additionalText => hashCode = hashCode * 397 ^ additionalText.GetHashCode());
```
Are you sure that we don't need the parentheses here?

FranckSix replied:
In C#, the * and ^ operators have the same precedence, and evaluation is performed from left to right.

Therefore, there is no difference. However, for readability, I have added the parentheses.

[1d13c38](https://github.com/ycanardeau/ResXGenerator/pull/7/commits/1d13c38fe63fabdb68752499fc8489b474053778)

ycanardeau reviewed (commented):
This PR is quite large, so it will take some time to review. It would be helpful if you could split it into multiple smaller parts.

ycanardeau commented on Aigamo.ResXGenerator.Tests/Aigamo.ResXGenerator.Tests.csproj:
```diff
@@ -33,11 +34,14 @@
 			<DependentUpon>$([System.IO.Path]::GetFileNameWithoutExtension('%(FileName)')).resx</DependentUpon>
 		</EmbeddedResource>
 		<EmbeddedResource Update="IntegrationTests\Test2.resx">
-			<GenerateCode>true</GenerateCode>
+			<GenerationType>CodeGeneration</GenerationType>
 		</EmbeddedResource>
```
I'm wondering if we should add a new test file rather than changing the existing one to ensure backward compatibility.

FranckSix replied:
Only to avoid [CS8602](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/nullable-warnings?f1url=%3FappId%3Droslyn%26k%3Dk(CS8602)#possible-dereference-of-null)

<img width="522" height="136" alt="image" src="https://github.com/user-attachments/assets/67771309-085f-4f31-a8b8-409866337811" />

Because of condition :
```cs
if (languageName.IsNullOrEmpty())
{
    return false;
}
```
I am pretty sure that languageName is not null.

FranckSix replied:
Reverted to original code for Test2.resx to GenerateCode, and I added a Test2a.resx to test CodeGeneration. It a good point because it permit to compare the two generated file and ensure they are identical.

[9318815](https://github.com/ycanardeau/ResXGenerator/pull/7/commits/931881533569242c02db60bea97b66059f36b2b2)

ycanardeau commented on Aigamo.ResXGenerator/Tools/Utilities.cs:
```diff
@@ -17,7 +21,7 @@ private static bool IsValidLanguageName(string? languageName)
 			}
 
-			if (languageName.StartsWith("qps-", StringComparison.Ordinal))
+			if (languageName!.StartsWith("qps-", StringComparison.Ordinal))
```
Why do you need the bang operator here?

FranckSix replied:
Only to avoid [CS8602](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/nullable-warnings?f1url=%3FappId%3Droslyn%26k%3Dk(CS8602)#possible-dereference-of-null)

<img width="522" height="136" alt="image" src="https://github.com/user-attachments/assets/24b8897b-dbac-4c60-bda3-838dfdc2780e" />

Because of condition :
```cs
if (languageName.IsNullOrEmpty())
{
    return false;
}
```
I am pretty sure that languageName is not null.

ycanardeau commented on Aigamo.ResXGenerator/Tools/Utilities.cs:
```diff
@@ -84,7 +88,7 @@ public static string GetLocalNamespace(
 			if (!string.IsNullOrWhiteSpace(targetPath))
 			{
-				localNamespace = Path.GetDirectoryName(targetPath)
+				localNamespace = Path.GetDirectoryName(targetPath)!
```
and here

FranckSix replied:
I added this to prevent a warning message about a potential NullReferenceException:

<img width="434" height="100" alt="image" src="https://github.com/user-attachments/assets/fd24f48d-58fa-4f39-948f-288c024a44c7" />

It would probably be safer to check that the GetDirectoryName method does not return null, rather than using the "!" operator. The original code didn't perform this check, which could lead to a NullReferenceException.

Do you want I do the full check. Or its ok ?

ycanardeau commented on Aigamo.ResXGenerator/Models/GenFilesNamespace.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Aigamo.ResXGenerator.Models;
+
+public record GenFilesNamespace(string Namespace, ImmutableArray<GenFileOptions> Files)
```
Just out of curiosity, I'm wondering if there are any benefits to using immutable arrays.

FranckSix replied:
Not specifically. Just to stay as close as possible to the IncrementalValuesProvider.

ycanardeau commented on Aigamo.ResXGenerator/Extensions/EnumerableExtensions.cs:
```diff
@@ -0,0 +1,9 @@
+
+public static class EnumerableExtensions
+{
+	public static void ForEach<T>(this IEnumerable<T> col, Action<T> action)
+	{
+		foreach (var i in col) action(i);
+	}
```
Do you prefer functional programming?

FranckSix replied:
Yes because it permit Linq lambdas method and avoid using of continue and yield commands. Is more modulable and more easy for refactoring and extract methods.

```cs
myList.ForEach(i =>
{
    if (i.IsValid) return;
}
//instead of
foreach(var i in myList)
{
    if(i.IsValid) continue; //Not fully understand by majority of coder. It's consider in my business like a goto. Also called spaghetti code.
    yield return i.Value; //Artefact before linq
}
```

The "yield return" keyword is a programming feature that was introduced before the LINQ library.

It is generally better to use, for example:
```cs
myList.Select(s => s.Value);
```
rather than
```cs
foreach(var i in myList)
{
    yield return i.Value;
}
```

When you uses LINQ methods you also benefit indirectly from the code optimizations that Microsoft has implemented in the LINQ library.

ycanardeau commented on Aigamo.ResXGenerator/Extensions/BoolExtensions.cs:
```diff
@@ -0,0 +1,6 @@
+
+public static class BoolExtensions
+{
+	public static T InterpolateCondition<T>(this bool condition, T valueIfTrue, T valueIfFalse) => condition ? valueIfTrue : valueIfFalse;
```
I'm wondering why we need this extension method instead of a ternary operator.

FranckSix replied:
This is a workaround for a limitation of the C# language. As you may have noticed, I love using raw strings in C#. They allow us to use string interpolation and maintain the visual formatting of indentation in the code. However, string interpolation doesn't support using conditional statements on a single line because of the colon (:).

```cs
   $"{condition ? true : false}" // Syntax error on ":" The colon break interpolation
```

The InterpolateCondition method allows you to overcome this problem and take advantage of the best of both worlds.

FranckSix committed:
Add comment to the GenerationType enum

Add comment to enum GenerationType

FranckSix committed:
Typo in commented code

Correction of typo in documentation

FranckSix committed:
Add XML Documentation in GenerationType

Report the README to XML comment on code to Intellisense purpose

FranckSix committed:
Adding parenthesis for clarity

In C#, the * and ^ operators have the same precedence, and evaluation is performed from left to right. So it's not strictly necessary, but I'll add it for clarity.In C#, the * and ^ operators have the same precedence, and evaluation is performed from left to right. So it's not strictly necessary, but I'll add it for clarity.

FranckSix committed:
Add an isolated test for CodeGeneration

Also reverted the original comportment on Test2.resx

FranckSix committed:
Correction of somes glitch after using it in real environement

The global registration class generated even if no StringLocalizer was present
Pass the message of exception instead of exception in case of fatal error
Environement.NewLine not recommended for Analyser and Generator. I changed for a constant
Adjust the Unshipped.md
Add a QUICKSTART.md to have a global view of possibilities the option and simplify configuration for new and current users.

FranckSix commented:
Hi, @ycanardeau

Did you have time to spot last update, I applied your recommendations. Note that I create a nuget on my actual project. I use the ResourceManager and ILocalizer modes. At this time it seem to work perfectly. And I can now generate ResxManager without the problematic "System.Globalization.CultureInfo.CurrentUICulture.LCID" for RoslynAnalyzer. I adding changeset [1cb23f0](https://github.com/ycanardeau/ResXGenerator/pull/7/commits/1cb23f076cd32aaf64c34ee109bb886261e87e76) after using it in real environment and spot somes irritants.

I updated documentation to adding a quick reference to help for configuration.

Have a nice day, let me know if you want I adjust something else.

FranckSix commented:
Hi @ycanardeau,

It's been a while since I've heard anything. I'm currently running an unofficial Nuget package published on a local server. Eventually, I'll have to switch to an official Nuget package... do you understand my dilemma?

Is there still hope of seeing my code integrated, or are you suggesting me to start a new branch and publish another version of the library?

ycanardeau commented:
Thank you!

FranckSix commented:
Thank you too, don't hesitate if you have some ideas or suggestions I can Help. :-)
